File: Localization_For_LunarLander/app/src/main/java/com/example/localization_for_lunarlander/LanderCanvas.java


package com.example.localization_for_lunarlander;

import android.content.Context;
import android.graphics.*;
import android.util.*;
import android.view.*;
import java.util.*;
//import stanford.androidlib.*;
import stanford.androidlib.SimpleBitmap;
import stanford.androidlib.graphics.*;
import stanford.androidlib.util.*;

public class LanderCanvas extends GCanvas {
    // frames per second of animation
    private static final int FRAMES_PER_SECOND = 30;

    // maximum velocity at which the rocket can hit the ground without crashing
    private static final float MAX_SAFE_LANDING_VELOCITY = 7.0f;

    private static final float ASTEROID_VELOCITY = -12.0f;

    // downward acceleration due to gravity
    private static final float GRAVITY_ACCELERATION = .5f;

    // upward acceleration due to thrusters
    private static final float THRUST_ACCELERATION = -.3f;

    // private fields
    private GSprite rocket;
    private GSprite moonSurface;
    private ArrayList<GSprite> asteroids = new ArrayList<>();
    private GLabel scoreLabel;
    private int score = 0;
    private Bitmap rocketNonThrust;
    private ArrayList<Bitmap> rocketThrust;
    private boolean isThrusting = false;

    /*
     * Required auto-generated constructor.
     */
    public LanderCanvas(Context context, AttributeSet attrs) {
        super(context, attrs);
        // GSprite.setDebug(true);
    }

    /*
     * Called when the GCanvas is being created.
     * Sets up the sprite objects and adds them to the screen.
     */
    @Override
    public void init() {
        setBackgroundColor(GColor.BLACK);

        // set up score label to display player's points earned
        String scoreText = getResources().getString(
                R.string.score_text, 0);
        scoreLabel = new GLabel(scoreText);
        scoreLabel.setColor(GColor.WHITE);
        scoreLabel.setFontSize(40);
        scoreLabel.setY(10);
        scoreLabel.setRightX(getWidth() - 20);
        add(scoreLabel);

        moonSurface = new GSprite(SimpleBitmap.with(this)
                .scaleToFit(R.drawable.moonsurface, this));
        moonSurface.setBottomY(getHeight());
        moonSurface.setCollisionMarginTop(moonSurface.getHeight()/3);
        add(moonSurface);

        rocketThrust = new ArrayList<>();
        rocketThrust.add(SimpleBitmap.with(this)
                .scaleToHeight(R.drawable.rocketshipthrust1, getHeight()/6));
        rocketThrust.add(SimpleBitmap.with(this)
                .scaleToHeight(R.drawable.rocketshipthrust2, getHeight()/6));
        rocketThrust.add(SimpleBitmap.with(this)
                .scaleToHeight(R.drawable.rocketshipthrust3, getHeight()/6));
        rocketThrust.add(SimpleBitmap.with(this)
                .scaleToHeight(R.drawable.rocketshipthrust4, getHeight()/6));

        rocketNonThrust = SimpleBitmap.with(this)
                .scaleToHeight(R.drawable.rocketship1, getHeight()/6);
        rocket = new GSprite(rocketNonThrust);
        rocket.setCollisionMargin(rocket.getWidth() / 4);
        rocket.setVelocityY(2);
        rocket.setAccelerationY(GRAVITY_ACCELERATION);
        add(rocket);
    }

    /*
     * Called by the GCanvas internal animation loop each time the animation ticks,
     * 30 times per second.
     * Updates the sprites and checks for collisions.
     */
    @Override
    public void onAnimateTick() {
        super.onAnimateTick();

        if (rocket.isMoving()) {
            // add an asteroid to the screen once per second
            if (getAnimationTickCount() % 30 == 0) {
                randomAsteroid();
                score++;
                String scoreText = getResources().getString(
                        R.string.score_text,
                        score
                );
                scoreLabel.setText(scoreText);
            }

            // check for collisions
            if (rocket.collidesWith(moonSurface)) {
                rocket.stop();
            } else {
                for (GSprite asteroid : asteroids) {
                    if (rocket.collidesWith(asteroid)) {
                        rocket.stop();
                    }
                }
            }
        }
    }

    /*
     * Adds an asteroid to the screen at a random y location.
     */
    private void randomAsteroid() {
        int randomY = RandomGenerator.getInstance().nextInt(0, getHeight());
        int x = getWidth();
        GSprite asteroid = new GSprite(SimpleBitmap.with(this)
                        .scaleToWidth(R.drawable.asteroid1, getWidth() / 20));
        asteroid.setLocation(x, randomY);
        asteroid.setVelocityX(ASTEROID_VELOCITY);
        asteroids.add(asteroid);
        add(asteroid);
    }

    /*
     * Called when the user presses or releases their finger from the screen.
     * Causes the rocket to start or stop thrusting.
     */
    @Override
    public boolean onTouch(View v, MotionEvent event) {
        float x = event.getX();
        float y = event.getY();
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            // code to run when finger is pressed; begin thrusting
            isThrusting = true;
            rocket.setAccelerationY(THRUST_ACCELERATION);
            rocket.setBitmaps(rocketThrust);
            rocket.setFramesPerBitmap(5);
            rocket.setLoopBitmaps(true);
        } else {
            // lifted finger up; stop thrusting
            isThrusting = false;
            rocket.setAccelerationY(GRAVITY_ACCELERATION);
            rocket.setBitmap(rocketNonThrust);
        }

        return super.onTouch(v, event);
    }
    /*
     * Called when user clicks Play Game button.
     * Starts a new game.
     */
    public void startGame() {
        animate(FRAMES_PER_SECOND);
    }

    /*
     * Called when user clicks Stop Game button.
     * Ends the game.
     */
    public void stopGame() {
        animationStop();
    }
}

File: AndroidStudioProjects/LunarLander/app/src/main/java/com/example/lunarlander/LanderCanvas.kt
package com.example.lunarlander

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.AttributeSet
import android.view.MotionEvent
import stanford.androidlib.graphics.*
import java.util.*

class LanderCanvas(context: Context, attrs: AttributeSet)
    : GCanvas(context, attrs) {

    // static constants，share value, not copy anywhere
    companion object {
        private const val FRAMES_PER_SECOND = 30

        private const val MAX_SAFE_LANDING_VELOCITY = 7.0f

        private const val ASTEROID_VELOCITY = -12.0f

        private const val GRAVITY_ACCELERATION = .5f

        private const val THRUST_ACCELERATION = -.3f

        private val ASTEROID_SIZE = 20f
        private val ASTEROID_DX = 12f
    }

    private lateinit var rocket: GSprite
    private lateinit var moonSurface: GSprite
    private lateinit var rocketImage: Bitmap
    private var rocketImageThrust = ArrayList<Bitmap>()




    override fun init() {
//        GSprite.setDebug(true)
        backgroundColor = GColor.BLACK


        var moonSurfaceImage = BitmapFactory.decodeResource(resources,R.drawable.moonsurface)
        moonSurfaceImage = moonSurfaceImage.scaleToWidth(this.width.toFloat())
        moonSurface = GSprite(moonSurfaceImage)
        moonSurface.bottomY = height.toFloat()
        //第二版碰撞检测
        moonSurface.collisionMarginTop = moonSurface.height / 4f
        add(moonSurface)

        rocketImage = BitmapFactory.decodeResource(resources,R.drawable.rocket)
        rocketImage = rocketImage.scaleToWidth(width/4f)
        rocket = GSprite(rocketImage)
        rocket.accelerationY = GRAVITY_ACCELERATION
        add(rocket)

        for(i in 1 until 4){
            var rocketImageThrustOne = BitmapFactory.decodeResource(resources,R.drawable.rocket+i)
            rocketImageThrustOne = rocketImageThrustOne.scaleToWidth(width/4f)
            rocketImageThrust.add(rocketImageThrustOne)
        }

        setOnTouchListener{
            _, event ->
            handleTouchEvent(event)
            true
        }

    }

    private fun handleTouchEvent(event: MotionEvent){
//        val x = event.x
//        val y = event.y
        if(event.action == MotionEvent.ACTION_DOWN) {
            rocket.accelerationY = THRUST_ACCELERATION
            rocket.bitmaps = rocketImageThrust
            rocket.framesPerBitmap = FRAMES_PER_SECOND /3
        } else if (event.action == MotionEvent.ACTION_UP){
            rocket.accelerationY = GRAVITY_ACCELERATION
            rocket.bitmap = rocketImage
        }
    }


    private var frame = 0
    private var score = 0
    private val scoreLabel = GLabel()
    private fun tick(){
        rocket.update()
        updateAsteroid()


        if(++frame % 30 == 0) {
            ++score
            showScore(score.toString())
        }
        doCollision()
    }

    private fun showScore(str: String){
        scoreLabel.text = "Time：$str s"
        scoreLabel.color = GColor.RED
        scoreLabel.fontSize = 50f
        scoreLabel.x = width/2f
        scoreLabel.y = height/6f
        add(scoreLabel)
    }

    private fun doCollision(){
        if (rocket.collidesWith(moonSurface)){
            if(rocket.velocityY <= MAX_SAFE_LANDING_VELOCITY){
                showMessage("SAFE LANDING You Win!",width/6f,height/2f)
            } else {
                // you lose
                showMessage("TOO FAST LANDING You Lose!",width/6f,height/2f)
            }

            rocket.velocityY = 0f
            rocket.accelerationY = 0f
            animationStop()
        }

    }

    private fun showMessage(str: String, x: Float, y: Float){
        val message = GLabel(str,x,y)
        message.color = GColor.RED
        message.fontSize = 50f
        add(message)
    }


    private var ticks = 0
    fun startGame() {
        animate(FRAMES_PER_SECOND){
            tick()
        }
    }

    fun stopGame() {
        animationStop()
    }

    private fun makeAsteroid() {
        val x = width
        val y = Random().nextInt((height-ASTEROID_SIZE).toInt())
        val asteroidGOval = GOval(x.toFloat(),y.toFloat(),ASTEROID_SIZE,ASTEROID_SIZE)
        asteroidGOval.fillColor = GColor.GRAY
        add(asteroidGOval)
    }

    private fun updateAsteroid() {
        ticks++
        if (ticks == 30){
            makeAsteroid()
            ticks = 0
        }

        for (shape in this) {
            if (shape != rocket && shape != moonSurface && shape != rocketImageThrust && shape != scoreLabel) {
                shape.moveBy(-ASTEROID_DX, 0f)
                if (shape.leftX < 0) {
                    remove(shape)
                }
            }
        }

    }



}